content/reference-package-objects.md

```diff
@@ -0,0 +1,253 @@
+
+```scala
+package object foo extends _root_.foo.bar.Qux{
+  val bar = 1
+}
+```
+```scala
+package foo.bar
+class Qux
+```
+
+We need to move the `package foo` contents into `package foo2` to avoid conflicts with 
+`object foo`, and then we need to add back aliases to all the declarations in `foo2` to make
+them available in `foo`:
+
+```scala
+object foo extends foo2.bar.Qux{
+  val bar = 1
+  object bar{
+    type Qux = foo2.bar.Qux
+  }
+}
+```
+```scala
+package foo2.bar
+class Qux
+```
+
+Both of these workarounds are awkward and non-idiomatic, but are necessary due to current 
+limitations in referencing `package object`s directly
+
+Notably, normal `object`s are not a replacement for `package object`s, because only
+`package object`s allow the package contents to be defined in other files. Normal `object`s
+would require that the package contents be all defined in a single file in the `object` body,
+or scattered into other files as `trait`s in _different_ `package`s and mixed into the
+`object`, both of which are messy and sub-optimal.
+
+It's possible to have a convention _"the `object` named `foo` is always going to be the
+primary entrypoint for a package"_, but that is just a poor-man's `package object` with worse
+syntax and less standardization.
+
+## Implementation Alternatives
+
+* We could make `a.b` where `b` is a `package` refer to the entire `package b` namespace, not 
+  just the `package object`. This cannot in general work due to the JVM's _open packages_ and
+  separate compilation: while `package object`s can only exist in one file present in one 
+  compilation run, JVM `package`s can contain arbitrary sets of classes from different compilation
+  runs. Thus it is impossible in general to define a "complete" API for a JVM `package` for us to
+  generate an object to refer to.
+
+* Using Scala 3 [Top Level Definitions](https://docs.scala-lang.org/scala3/book/taste-toplevel-definitions.html)
+  is one possible alternative to `package object`s, but they fall short on many use cases:
+  * Top-level definitions cannot generate objects that inherit from classes or traits, which
+    is necessary in many use cases: Mill (needs them to inherit `mill.Module`), Requests-
+    Scala (needs it to inherit from `requests.BaseSession`), etc.
+  * Top-level definitions can be defined in multiple files, so suffer from the issue that 
+    it is at any point in time impossible to know the "entire" API of a `package` provided
+    by top-level definitions
+  * Top-level definitions do not provide a natural "package entrypoint" to the `package` source folder,
+    to provide package-level documentation, annotations, etc.. We could provide another `.scala` 
+    file that we specify by-convention to be the "package entrypoint", but we already have
+    `package.scala` and it does the job just fine
+
+
+## Limitations
+
+* `a.b` only expands to `a.b.package` when used "standalone", i.e. not when part of a
```

Comment:
I don't think this limitation is necessary and it would be hard to implement. I think we can simply say that in selection path `a.b.c`, `a.b` means the full package, not just the package object. But it's OK to have operators. By the time we check valid package object use, we have already expanded out all these operators to selections.